rllib/agents/dqn/r2d2_tf_policy.py

```diff
@@ -0,0 +1,346 @@
+    weights = tf.cast(train_batch[PRIO_WEIGHTS], tf.float32)
+
+    B = tf.shape(state_batches[0])[0]
+    T = tf.shape(q)[0] // B
+
+    # Q scores for actions which we know were selected in the given state.
+    one_hot_selection = tf.one_hot(actions, policy.action_space.n)
+    q_selected = tf.reduce_sum(
+        tf.where(q > tf.float32.min, q, 0.0) * one_hot_selection, axis=1)
+
+    if config["double_q"]:
+        best_actions = tf.argmax(q, axis=1)
+    else:
+        best_actions = tf.argmax(q_target, axis=1)
+
+    best_actions_one_hot = tf.one_hot(best_actions, policy.action_space.n)
+    q_target_best = tf.reduce_sum(
+        tf.where(q_target > tf.float32.min, q_target, 0.0) *
+        best_actions_one_hot,
+        axis=1)
+
+    if config["num_atoms"] > 1:
+        raise ValueError("Distributional R2D2 not supported yet!")
+    else:
+        q_target_best_masked_tp1 = (1.0 - dones) * tf.concat(
+            [q_target_best[1:], tf.constant([0.0])], axis=0)
+
+        if config["use_h_function"]:
+            h_inv = h_inverse(q_target_best_masked_tp1,
+                              config["h_function_epsilon"])
+            target = h_function(
+                rewards + config["gamma"]**config["n_step"] * h_inv,
+                config["h_function_epsilon"])
+        else:
+            target = rewards + \
+                config["gamma"] ** config["n_step"] * q_target_best_masked_tp1
+
+        # Seq-mask all loss-related terms.
+        seq_mask = tf.sequence_mask(train_batch["seq_lens"], T)[:, :-1]
+        # Mask away also the burn-in sequence at the beginning.
+        burn_in = policy.config["burn_in"]
+        # Making sure, this works for both static graph and eager.
+        if burn_in > 0 and (config["framework"] == "tf" or burn_in < T):
+            seq_mask = tf.concat(
+                [tf.fill([B, burn_in], False), seq_mask[:, burn_in:]], axis=1)
+
+        def reduce_mean_valid(t):
+            return tf.reduce_mean(tf.boolean_mask(t, seq_mask))
+
+        # Make sure use the correct time indices:
+        # Q(t) - [gamma * r + Q^(t+1)]
+        q_selected = tf.reshape(q_selected, [B, T])[:, :-1]
+        td_error = q_selected - tf.stop_gradient(
+            tf.reshape(target, [B, T])[:, :-1])
+        td_error = td_error * tf.cast(seq_mask, tf.float32)
+        weights = tf.reshape(weights, [B, T])[:, :-1]
+        policy._total_loss = reduce_mean_valid(weights * huber_loss(td_error))
+        policy._td_error = tf.reshape(td_error, [-1])
+        policy._loss_stats = {
+            "mean_q": reduce_mean_valid(q_selected),
+            "min_q": tf.reduce_min(q_selected),
+            "max_q": tf.reduce_max(q_selected),
+            "mean_td_error": reduce_mean_valid(td_error),
+        }
+
+    return policy._total_loss
+
+
+def h_function(x, epsilon=1.0):
+    """h-function to normalize target Qs, described in the paper [1].
+
+    h(x) = sign(x) * [sqrt(abs(x) + 1) - 1] + epsilon * x
```

Comment:
Perhaps put in a utils file.

Reply:
I would, but what should we call this function? It's not named in the paper either (except for being the "h" function). If I leave it in R2D2, everyone knows, it's a function from the paper. If I put it into utils, it's not clear what other use this function would have outside of R2D2 and what we should name it.
It's similar to a tanh/sigmoid, but without any bounds.

Reply:
Aight its fine to keep it in the policy file